### PR TITLE
8299677: Formatter.format might take a long time to format an integer or floating-point

### PR DESCRIPTION
**Pull request by rgiulietti**
This change transforms a O(n^2) path to O(n) when prepending zero padding to decimal outputs, where n is the length of the padding.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8299677](https://bugs.openjdk.org/browse/JDK-8299677): Formatter.format might take a long time to format an integer or floating-point

### Reviewers
 * [Alan Bateman](https://openjdk.org/census#alanb) (@AlanBateman - **Reviewer**) ⚠️ Review applies to [275582a9](https://git.openjdk.org/jdk/pull/11939/files/275582a9860c7da9deeb53a979109eaa304f6deb)
 * [Sergey Tsypanov](https://openjdk.org/census#stsypanov) (@stsypanov - Author)
 * [Joe Darcy](https://openjdk.org/census#darcy) (@jddarcy - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11939/head:pull/11939` \
`$ git checkout pull/11939`

Update a local copy of the PR: \
`$ git checkout pull/11939` \
`$ git pull https://git.openjdk.org/jdk pull/11939/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11939`

View PR using the GUI difftool: \
`$ git pr show -t 11939`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11939.diff">https://git.openjdk.org/jdk/pull/11939.diff</a>

</details>

### CODE REVIEW

**Comment by AlanBateman** [comment]
I assume this issue has been there for a long time, just not noticed because it would be unusual to specify a width of this magnitude. The switch to using repeat + one insert looks good. Are you okay for tests in this area?

**Comment by rgiulietti** [comment]
@AlanBateman This is not a functional bug, only an enhancement to overcome the long times for large, unrealistic widths.
Consequently, I have no specific tests that measure the timing against some hard-coded limits, if I understand you correctly.

**Comment by AlanBateman** [comment]
> @AlanBateman This is not a functional bug, only an enhancement to overcome the long times for large, unrealistic widths.
> Consequently, I have no specific tests that measure the timing against some hard-coded limits, if I understand you correctly.

My question wasn't phrased very well, I was really just asking if we have good test coverage for width in the format specifier  so that it would catch any regressions from changes to this code.

**Comment by rgiulietti** [comment]
Since specific tests for padding seem lacking, I'll add some as part of this PR.
Please expect another commit before marking as reviewed.

**Comment by rgiulietti** [comment]
Added straightforward tests for padding.

File: src/java.base/share/classes/java/util/Formatter.java [hunk 1]
```diff
@@ -4596,10 +4596,9 @@ private StringBuilder localizedMagnitude(Formatter fmt, StringBuilder sb,
             }
 
             // apply zero padding
-            if (width != -1 && Flags.contains(f, Flags.ZERO_PAD)) {
-                for (int k = sb.length(); k < width; k++) {
-                    sb.insert(begin, zero);
-                }
+            if (width > sb.length() && Flags.contains(f, Flags.ZERO_PAD)) {
+                String zeros = Character.toString(zero).repeat(width - sb.length());
```
Comment (stsypanov) [inline src/java.base/share/classes/java/util/Formatter.java:4600]: `Character.toString(zero)` could be replaced with `String.valueOf(zero)` saving one more method call

File: test/jdk/java/util/Formatter/Padding.java [hunk 2]
```diff
@@ -0,0 +1,362 @@
...
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+/*
+ * @test
+ * @run junit Padding
```
Comment (AlanBateman) [inline test/jdk/java/util/Formatter/Padding.java:26]: Probably should add a @summary tag to this, maybe a @bug too (I think JDK-4906370 added Formatter in Java 5).

**Review by AlanBateman — APPROVED** [review]
Thanks for expanding the tests for this.

**Review by LanceAndersen — COMMENTED** [review]
Overall this looks OK.  

Please add either a general comment describing what we are trying to test or a comment for each test method.  As we add new tests,  adding comments for additional clarity will benefit future maintainers

File: test/jdk/java/util/Formatter/Padding.java [hunk 3]
```diff
@@ -0,0 +1,362 @@
...
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+
+public class Padding {
+
+    @Nested
+    class BlankPaddingInt {
+
+        @Test
```
Comment (LanceAndersen) [inline test/jdk/java/util/Formatter/Padding.java:39]: Any thought given to using `@MethodSource:` in order to specify a DataProvider? The benefit would be you could catch each failure vs having to address or comment out the assertEquals that is failing

**Comment by rgiulietti** [comment]
All concerns were addressed, I hope.

**Review by stsypanov — APPROVED** [review]

**Review by jddarcy — APPROVED** [review]

**Review by balcanuc — COMMENTED** [review]
in which JDK versions will the bug-fix back ported? As it is, it is impossible to be back ported to JDK8

**Comment by rgiulietti** [comment]
@balcanuc Backporting is done in other, separate projects. Whether changes in mainline are backported, and how, is publicly discussed in the respective mailing lists. 

In particular, for JDK 8 Updates you might want to have a look at
[this mailing list](https://mail.openjdk.org/mailman/listinfo/jdk8u-dev) and the [this git repo](https://github.com/openjdk/jdk8u-dev).

If you feel this change is relevant to JDK 8, you are welcome to discuss a solution that does not rely on `repeat()` on that mailing list, and then [contribute](https://openjdk.org/contribute/) your code to the [JDK 8 Update project](https://openjdk.org/projects/jdk8u/).

**Pull request closed without merging**